### PR TITLE
SONAR-6494 WS permissions/apply_template apply a permission template to a project

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionService.java [hunk 1]
```diff
@@ -160,40 +163,34 @@ public void applyDefaultPermissionTemplate(final String componentKey) {
     indexProjectPermissions();
   }
 
+  /**
+   * @deprecated since 5.2 – to be deleted once Permission Template page does not rely on Ruby
+   */
+  @Deprecated
   public void applyPermissionTemplate(Map<String, Object> params) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionService.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/WsProjectRef.java [hunk 2]
```diff
@@ -21,49 +21,54 @@
...
  * Project identifiers from a WS request. Guaranties the project id and project key are not provided at the same time.
  */
-class WsProject {
+class WsProjectRef {
   private final String uuid;
   private final String key;
 
-  private WsProject(String uuid, String key) {
+  private WsProjectRef(Request wsRequest) {
+    String uuid = wsRequest.param(PARAM_PROJECT_ID);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/WsProjectRef.java]: MAJOR Rename "uuid" which hides the field declared at line 35. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/WsProjectRef.java [hunk 3]
```diff
@@ -21,49 +21,54 @@
...
+    String key = wsRequest.param(PARAM_PROJECT_KEY);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/WsProjectRef.java]: MAJOR Rename "key" which hides the field declared at line 36. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionRepository.java [hunk 4]
```diff
@@ -134,17 +134,19 @@ private void updateProjectAuthorizationDate(DbSession session, @Nullable Long pr
...
   }
 
-  public void applyPermissionTemplate(DbSession session, String templateKey, Long resourceId) {
-    PermissionTemplateDto permissionTemplate = dbClient.permissionTemplateDao().selectPermissionTemplateWithPermissions(session, templateKey);
+  public void applyPermissionTemplate(DbSession session, String templateUuid, long resourceId) {
+    PermissionTemplateDto permissionTemplate = dbClient.permissionTemplateDao().selectPermissionTemplateWithPermissions(session, templateUuid);
     updateProjectAuthorizationDate(session, resourceId);
     dbClient.roleDao().removeAllPermissions(session, resourceId);
     List<PermissionTemplateUserDto> usersPermissions = permissionTemplate.getUsersPermissions();
+    //TODO should return an empty list if there's no user permissions
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionRepository.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionRepository.java [hunk 5]
```diff
@@ -134,17 +134,19 @@ private void updateProjectAuthorizationDate(DbSession session, @Nullable Long pr
...
     if (usersPermissions != null) {
       for (PermissionTemplateUserDto userPermission : usersPermissions) {
         insertUserPermission(resourceId, userPermission.getUserId(), userPermission.getPermission(), false, session);
       }
     }
     List<PermissionTemplateGroupDto> groupsPermissions = permissionTemplate.getGroupsPermissions();
+    //TODO should return an empty list if there's no group permission
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionRepository.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 2 major
* INFO 3 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* INFO 3 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* INFO 3 info

Watch the comments in this conversation to review them.

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionTemplateDto.java [hunk 6]
```diff
@@ -66,11 +66,26 @@ public String getKee() {
     return kee;
   }
 
+  /**
+   * @since 5.2 the kee column is a proper uuid. Before that it was build on the name + timestamp
+   */
+  public String getUuid() {
```
Comment (simonbrandhof) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionTemplateDto.java]: As discussed the db column must be renamed in order to remove the ambiguity. Key and uuid in the DTO are the same.

**Comment by simonbrandhof** [comment]
As discussed, we'll do a global review of permission codebase after removal of ruby on rails webapp. There is too much java code that brings noise for review.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* INFO 3 info

Watch the comments in this conversation to review them.

**Pull request closed without merging**